# Convex Optimization: Linear Programs

**Prerequisites**

- Linear Algebra
- Calculus
- Convex Optimization: Theoretical Foundations


**Outcomes**

- Understand the format of a linear program
- Know how to formulate a linear program in standard form
- Understand and interpret slack variables
- Know how to set up and solve linear programs

**Outline**

- Linear Programs
- Example: Diet Problem
- Optimal Transport Primer
- Example: Product Mix Problem

In [1]:
import numpy as np

## Linear Programs

A **linear program** is an optimization problem of the form

$$\begin{align*}
\min_x \ & c^T x \\
\mbox{subject to } \ & Gx \le h \\
& Ax = b\\
\end{align*}$$

### Standard Form

A *linear program in standard form* has only equality constraints and a non-negativity constraint on $x$:

$$\begin{align*}
\min_x \ & c^T x \\
& Ax = b,\\
& x_i \ge 0 \; \forall i
\end{align*}$$

<a id='exercise-0'></a>
**Exercise 1**

Convert a linear program in the general form to a linear program in standard form


*Hint:* You will need to introduce a new vector (of slack variables) into the constraints

## Example: Diet Problem

**Want**: Construct calorie minimizing diet of $m$ distinct foods, subject to obtaining a minimum level of $k$ nutrients

Let 

- $c_i$ for $i = 1, \dots, m$ represent calories per serving of food $i$ 
- $a_{i,j}$ for $i=1,\dots,k$ and $j=1,\dots, m$ represent the amount of nutrient $i$ in food $j$
- $b_i$ for $i=1,\dots,k$ be the minimum level of nutrient $i$ needed






##### A linear program

We can formulate the diet problem as a linear program by

- Constructing a matrix $A$ of all the $a_{k,m}$ values. Each column will represent nutrient values per serving of a single food, while each row will be value of a single nutrient in all foods
- Stack the nutrient requirements $b_k$ in a column vector, matching the order of the rows of $A$

Then, the linear program can be written

$$\begin{align*}
\min_x \ & c^T x \\
& Ax \ge b,\\
& x_i \ge 0 \; i=1,\dots,m
\end{align*}$$

> NOTE: the line $Ax \ge b$ is shorthand for imposing the inequality elementwise (row by row)

### Convert to standard form

We can convert the linear program to standard form by:

- Multiplying the $Ax \ge b$ inequality by $-1$, which is defining $\tilde{A} = -A$ and $\tilde{b} = -b$. Now inequality constraint is $$\tilde{A} x \le \tilde{b}$$
- Add a vector slack variables to the modified inequality: $$\tilde{A} x + s = \tilde{b}$$
- Impose non-negativity on our slack vairables

The standard form LP is 

$$\begin{align*}
\min_x \ & c^T x \\
& \tilde{A}x  + s \le \tilde{b},\\
& x_i \ge 0 \; i=1,\dots,m\\
& s_i \ge 0 \; i=1,\dots,k
\end{align*}$$

## Optimal Transport Primer

<a id='exercise-1'></a>
**Exercise 2: Transportation problem**

> Note: This problem comes from "Labs for Foundations of Applied Mathematics Volume 2" by Jeffrey Humpherys And Tyler J. Jarvis released under the [Creative Commons Attribution 3.0 United States license](http://creativecommons.org/licenses/by/3.0/us/). The original source for the lab materials can be found at [https://github.com/Foundations-of-Applied-Mathematics/Labs](https://github.com/Foundations-of-Applied-Mathematics/Labs)

Consider the following transportation problem: 

- A piano company needs to transport thirteen pianos from their three supply centers (denoted by 1, 2, 3) to two demand centers (4, 5)
- Transporting a piano from a supply center to a demand center incurs a cost (see table below)
- The company wants to minimize shipping costs for the pianos while meeting the demand

| Supply Center | Demand Center | Cost of Transportation | Number of Pianos | 
| :-----------: | :-----------: | :--------------------: | :--------------: |
| 1 | 4 | 4 | $p_1$ |
| 1 | 5 | 7 | $p_2$ | 
| 2 | 4 | 6 | $p_3$ | 
| 2 | 5 | 8 | $p_4$ |
| 3 | 4 | 8 | $p_5$ |
| 3 | 5 | 9 | $p_6$ |

The number of pianos available at each supply center is given by:

| Supply Center | Number of pianos available |
| :-----------: | :------------------------: |
| 1 | 7 |
| 2 | 2 | 
| 3 | 4 |

The number of pianos needed at each demand center is given by

| Demand Center | Number of pianos needed |
| :-----------: | :---------------------: |
| 4 | 5 |
| 5 | 8 |

A linear program can be set up to solve for $p_i$, $i=1,\dots,6$

The objective is to minimize total cost of shipping pianos, subject to these constraints:

- There cannot be a negative number of pianos transported along any route
- All supply centers must transport all pianos
- All demand centers must end up with needed number of pianosFinally, the objective function is the number of pianos shipped along each
route multiplied costs found in the costs table.

Your task is to formulate the transportation problem as a linear program

That is define the vector $c$, matrix $A$ and vector $b$ that appropriately defines the problem as described above

*Hint:* the matrix $A$ will have 5 rows and 6 columns

*your work here*

## Example: Product Mix Problem

> Note: This problem comes from chapter 13 of "Labs for Foundations of Applied Mathematics Volume 2" by Jeffrey Humpherys And Tyler J. Jarvis released under the [Creative Commons Attribution 3.0 United States license](http://creativecommons.org/licenses/by/3.0/us/). The original source for the lab materials can be found at [https://github.com/Foundations-of-Applied-Mathematics/Labs](https://github.com/Foundations-of-Applied-Mathematics/Labs)

Suppose...

- We work for a manufacturing plant 
- Our firm produces multiple goods, indexed by $i$
- Goods produced with common resources, indexed by $j$
- For next period, we know:
    - Maximum consumer demand for each good: $d_i$
    - The market price for each good: $p_i$
    - Amount of each resource our firm has: $m_j$
    - Amount of resource $j$ needed to produce  one unit of good $i$: $h_{ij}$
- We must choose amount of each good ($x_i$) to produce to maximize next period revenue

> Note: We acknowledge that this is a simplified view of an economy. This is intentional as we want to focus on the ability to cast this problem as a linear program.

### Problem setup

Our firm's revenue in the next period is given by 

$$\sum_i x_i p_i$$

The resource constraints can be expressed as

$$\sum_{i=1} h_{ji} x_i \le m_j \; \forall j$$

Finally, we must express the constraint that we cannot sell more than the consumers are willing to buy:

$$x_i \le d_i \; \forall i$$

We can cast this as a standard linear program (as described above) by setting:

\begin{align*}
c^T &= - \begin{bmatrix} p_1 & \cdots & p_n \end{bmatrix} \\
A &= \begin{bmatrix} h_{11} & \cdots & h_{1n} \\ \vdots & \ddots & \vdots \\ h_{j1} & \cdots & h_{jn} \\  &  \\ & I_n \\ & \end{bmatrix} \\
b &= \begin{bmatrix} m_1 & \cdots & m_j & d_1 & \cdots & d_n \end{bmatrix}^T \\
x_i &\ge 0 \; \forall i,
\end{align*}

where $I_n$ represents the  $n \times n$ identity matrix

The linear program is then:

$$\begin{align*}
\min_x \ & c^T x \\
\mbox{subject to } \ & Ax \le b \\
& x \ge 0\\
\end{align*}$$


> Note we would need to introduce slack variables to have only a single inequality constraint on $x$, but as you'll see below we can work with the problem as we just wrote it

### Solving the Product Mix Problem

Let's use the `scipy.optimize.linprog` routine to solve a version of the product mix problem

Below we have written out numerical values for the problem described above:

In [2]:
p = np.array([250., 215., 275., 180.])  # price vector
d = np.array([10., 20., 12., 10.])      # demand vector
m = np.array([4., 4., 4.])              # resource constraints

H = np.array([                          # resource usage
    [0.12, 0.18, 0.13, 0.07],
    [0.15, 0.12, 0.13, 0.11],
    [0.1 , 0.1 , 0.1 , 0.12]
])

#### To General Form

We apply the transformation above to set up a standard linear program

In [3]:
c = -p
b = np.concatenate((m, d))
A = np.vstack((H, np.eye(len(p))))

We will use the `scipy.optimize.linprog` routine to solve our linear program

Let's study its docstring:


In [6]:
from scipy.optimize import linprog
linprog?

Signature:
linprog(
    c,
    A_ub=None,
    b_ub=None,
    A_eq=None,
    b_eq=None,
    bounds=None,
    method='interior-point',
    callback=None,
    options=None,
    x0=None,
)
Docstring:
Linear programming: minimize a linear objective function subject to linear
equality and inequality constraints.

Linear programming solves problems of the following form:

.. math::

    \min_x \ & c^T x \\
    \mbox{such that} \ & A_{ub} x \leq b_{ub},\\
    & A_{eq} x = b_{eq},\\
    & l \leq x \leq u ,

where :math:`x` is a vector of decision variables; :math:`c`,
:math:`b_{ub}`, :math:`b_{eq}`, :math:`l`, and :math:`u` are vectors; and
:math:`A_{ub}` and :math:`A_{eq}` are matrices.

Informally, that's:

minimize::

    c @ x

such that::

    A_ub @ x <= b_ub
    A_eq @ x == b_eq
    lb <= x <= ub

Note that by default ``lb = 0`` and ``ub = None`` unless specified with
``bounds``.

Parameters
----------
c : 1-D array
    The coefficients of the linear objective function to be minimized.
A

In [7]:
linprog(c=c, A_ub=A, b_ub=b)

     con: array([], dtype=float64)
     fun: -7453.596491152559
 message: 'Optimization terminated successfully.'
     nit: 10
   slack: array([5.52287105e-11, 3.70965481e-11, 9.65964912e-01, 1.68914482e-09,
       1.38070175e+01, 1.65481850e-10, 8.21052631e+00])
  status: 0
 success: True
       x: array([10.        ,  6.19298246, 12.        ,  1.78947369])

From the above we can see the following:

- Total revenue will be \$7,453.59
- The optimal amount of the goods is approximately (10, 6.2, 12, 1.8)
- Given the values of the slack variables, we see that we:
    - Used all of resources 1 and 2, 
    - ... but had 0.96 units of resource 3 left
    - Satisfied all of the demand for goods 1 and 3
    - ... but were 13.8 units shy of demand for good 2
    - ... and 8.21 units shy of demand for good 4

<a id='exercise-2'></a>
**Exercise 3: Solving the Transportation Problem**

Using `scipy.optimize.linprog` solve the transportation problem from exercise 2 above.

In [ ]:
def solve_transportation():
    # your code here
    
    # define c
    
    # define A
    
    # define b
    
    # call linprog
    
    # return result of linprog
    
    pass  # delete this line when finished

solve_transportation()

### References

Boyd, Stephen, Stephen P. Boyd, and Lieven Vandenberghe. 2004. Convex Optimization. Cambridge University Press.

Jeffrey Humpherys And Tyler J. Jarvis. "Labs for Foundations of Applied Mathematics Volume 2"